design/progress-tracking.md

```diff
@@ -0,0 +1,244 @@
+phase's subsection as `status.actions[*].phases[*].progess`.
+
+For example,
+
+```yaml
+status:
+  progress:
+    percentCompleted: 50 # 1 out of 2 phases are completed
+    lastTransitionTime: 2022-04-06 14:23:34
+  actions:
+  - blueprint: my-blueprint
+    name: action-00
+    phases:
+    - name: echo
+      state: completed
+      progress:
+        percentCompleted: 100
+        lastTransitionTime: 2022-04-06 14:13:00
+    - name: echo
+      state: completed
+      progress:
+        percentCompleted: 100
+        lastTransitionTime: 2022-04-06 14:23:34
+    name: action-01
+    phases:
+    - name: echo
+      state: pending
+      progress:
+        percentCompleted: 30
+        lastTransitionTime: 2022-04-06 14:30:31
+  state: pending
+```
+
+### Progress Tracking
+
+Since progress tracking may not be meaningful for short-lived tasks, we will
+limit the initial implementation effort to the following Kanister Functions
+which normally are used to invoke long-running operations:
+
+* `BackupData`
+* `BackupDataAll`
+* `RestoreData`
+* `RestoreDataAll`
+* `CopyVolumeData`
+* `CreateRDSSnapshot`
+* `ExportRDSSnapshotToLocation`
+* `RestoreRDSSnapshot`
+
+#### Blueprint Progress
+
+Initially, the blueprint overall progress is computed by checking the number of
+completed phases against the total number of phases:
+
+```
+blueprint_progress_percent = num_completed_phases / total_num_phases * 100
+```
+
+In subsequent implementation, the computation alogrithm can be updated to assign
+more weight to phases with long-running operations. It's also possible to post
+periodic progress updates using an exponential backoff mechanism as long as the
+underlying phases are still alive.
+
+When the first phase of the blueprint is started, the blueprint progress will be
+updated to 10%, instead of keeping it at 0%. This will help to distinguish
+between in-progress first phase from an inactive first phase.
+
+The blueprint progress will only be set to 100% after all the phases completed
+without failures. The blueprint progress should not exceed 100%.
+
+#### Phase Progress
+
+As each phase within a blueprint may involve executing different commands
+producing different outputs, this design proposes a phase progress tracking
+interface that can use different "trackers" to map command outputs to numeric
+progress status.
+
+Some example trackers include ones that can track progress by:
+
+* checking the number of uploaded bytes against estimated total bytes
+* checking the duration elapsed against the estimated duration to complete the
+operation
+* parsing the log outputs for milestones indicator
+
+✍️ Currently, Kanister Functions do not utilize Kopia to perform their
+underlying work. Once the work to integrate Kopia into Kanister is completed,
+we can extract the progress status directly from the log outputs.
+
+Here's a sample log output of the Kopia create snapshot function:
+
+```sh
+$ kopia snapshot create kanister
```

Comment:
In both the examples (kopia and restic) we are just considering the utilities that are going to upload the data to object store but in some cases there are chances that there are some other commands that are taking some time.
- One example is the utility that we are using to take backup for example `mysqldump`, `esdump` etc. Should we consider these as well or just measuring upload time would be sufficient.

Reply:
Definitely if we can get progress from the tools we should expose that.

Reply:
@viveksinghggits Yes, the goal is that downstream users can add more custom progress trackers (at the code level) per their requirements, if the built-in ones don't work. We can consider adding a built-in generic tracker to output to 25%, 50%, 75% and 100% markers based on certain events in the logs.